### PR TITLE
Create initial version of Guardian Weekly holiday-stop-processor lambda

### PR DESCRIPTION
**Pull request by kelvin-chappell**
This is a lambda that covers the simplest possible case and just outputs the response from the subscription update if it's successful, or an error message if there's a failure.

I still have to get some environment variables from S3.

I've included a [standalone app](handlers/zuora-gw-holiday-stop/src/main/scala/com/gu/zuoragwholidaystop/StandaloneApp.scala) to make it easy to test end to end. 

The [lambda](handlers/zuora-gw-holiday-stop/src/main/scala/com/gu/zuoragwholidaystop/Handler.scala) and the standalone app are thin wrappers around the [essential process](handlers/zuora-gw-holiday-stop/src/main/scala/com/gu/zuoragwholidaystop/HolidayStopProcess.scala).

/cc @paulbrown1982 @mario-galic @twrichards

### CODE REVIEW

File: handlers/zuora-gw-holiday-stop/src/main/scala/com/gu/zuoragwholidaystop/Subscription.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    ratePlanCharges: Seq[RatePlanCharge]
+)
+
+case class RatePlanCharge(price: Double, billingPeriod: String) {
+
+  val weekCount: Int = {
+    // inexact values
+    billingPeriod match {
+      case "Month"   => 4
+      case "Quarter" => 12
```
Comment (paulbrown1982) [inline handlers/zuora-gw-holiday-stop/src/main/scala/com/gu/zuoragwholidaystop/Subscription.scala]: A quarter is mostly 13 weeks (hence that x 4 = 52).

File: handlers/zuora-gw-holiday-stop/src/main/scala/com/gu/zuoragwholidaystop/SubscriptionUpdate.scala [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      stoppedPublicationDate: LocalDate
+  ): SubscriptionUpdate = {
+    val credit =
+      autoRenewingHolidayCredit(subscription)
+    SubscriptionUpdate(
+      Seq(
+        Add(
+          productRatePlanId,
+          contractEffectiveDate = stoppedPublicationDate,
+          customerAcceptanceDate = stoppedPublicationDate,
```
Comment (paulbrown1982) [inline handlers/zuora-gw-holiday-stop/src/main/scala/com/gu/zuoragwholidaystop/SubscriptionUpdate.scala]: This would need to align to an invoicing date, else it will generate a dangling negative invoice (which does have a background job to clear it up, but that's an awkward project to maintain as it has to use the SOAP API to create credit balance adjustments).

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  def apply(): Either[String, Config] = {
+    val stage = Option(System.getenv("Stage")).getOrElse("DEV")
+    val s3Config = configFromS3(stage)
+    stage match {
+      case "PROD" =>
+        s3Config map { secretConfig =>
+          Config(
+            secretConfig,
+            holidayCreditProductRatePlanId = "2c92a0fc5b42d2c9015b6259f7f40040",
+            // todo
```
Comment (twrichards) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: what is there TODO?

Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: Well spotted!  That's a placeholder because I had difficulty creating the new one-off charge.  I'll fill in the value before I merge.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  zuoraAccess: ZuoraAccess,
+  holidayCreditProductRatePlanId: String,
+  holidayCreditProductRatePlanChargeId: String
+)
+
+case class ZuoraAccess(baseUrl: String, username: String, password: String)
+
+object Config {
+
+  private def configFromS3(stage: String): Either[String, ZuoraAccess] = {
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: Config errors are probably unrecoverable, so personally, I don't model them and just throw fast.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    val profileName = "membership"
+    val bucketName = "gu-reader-revenue-private"
+    val key =
+      if (stage == "DEV")
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: Can DEV config be renamed to match PROD such that instead of switching on stage we write 

```
val stage = getStage
val key = s"membership/support-service-lambdas/$stage/zuoraRest-$stage.v1.json"
```

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.gu.holidaystopprocessor
+
+import scala.math.BigDecimal.RoundingMode
+
+object Credit {
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala]: Instead of `Credit.autoRenewingHolidayAmount(sub)` we could write `AutoRenewingHolidayAmount(sub)` if we provide the following definition

```
object AutoRenewingHolidayAmount { def apply(sub) ... }
```

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  customerAcceptanceDate: LocalDate,
+  serviceActivationDate: LocalDate,
+  chargeOverrides: Seq[ChargeOverride]
+)
+
+case class ChargeOverride(
+  productRatePlanChargeId: String,
+  HolidayStart__c: LocalDate,
+  HolidayEnd__c: LocalDate,
+  price: Double
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: From my understanding `price` should be negative so could we add `require(price <= 0)` in constructor.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+  def autoRenewingHolidayAmount(subscription: Subscription): Double = {
+    def roundUp(d: Double): Double =
+      BigDecimal(d).setScale(2, RoundingMode.UP).toDouble
+    subscription.originalRatePlanCharge map { charge =>
+      val recurringPayment = charge.price
+      val numPublicationsInPeriod = charge.weekCount
+      -roundUp(recurringPayment / numPublicationsInPeriod)
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala]: Negative result seems to be an invariant here, se would could assert on this. See comment on `ChargeOverride` construction.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package com.gu.holidaystopprocessor
+
+object HolidayStopProcess {
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala]: How about rename to `ProcessHolidayStop` as it is an action.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+    val subscriptionDetails =
+      Zuora.subscriptionGetResponse(config.zuoraAccess) _
+
+    val updatedSubscription =
+      Zuora.subscriptionUpdateResponse(config.zuoraAccess) _
+
+    val holidayCreditToAdd = SubscriptionUpdate.holidayCreditToAdd(config) _
+
+    subscriptionDetails(stop.subscriptionName).right flatMap { subscription =>
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala]: Why `.right`? Eithers are already right-biased in Scala 2.12.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/StandaloneApp.scala [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package com.gu.holidaystopprocessor
+
+import java.time.LocalDate
+
+object StandaloneApp extends App {
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/StandaloneApp.scala]: Maybe document this is for testing purposes only.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Subscription.scala [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  ratePlanCharges: Seq[RatePlanCharge]
+)
+
+case class RatePlanCharge(
+  price: Double,
+  billingPeriod: Option[String],
+  effectiveEndDate: LocalDate
+) {
+  val weekCount: Int = {
+    // inexact values
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Subscription.scala]: Maybe move the comment in the name itself like `weekCountApprox`. It is ugly but it probably should be to provide signal to users.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+  def holidayCreditToAdd(config: Config)(
+    subscription: Subscription,
+    stoppedPublicationDate: LocalDate
+  ): SubscriptionUpdate = {
+    val credit = Credit.autoRenewingHolidayAmount(subscription)
+    val effectiveDate = subscription.originalRatePlanCharge map {
+      _.effectiveEndDate.plusDays(1)
+    } getOrElse stoppedPublicationDate
+    val currentTerm =
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: Factor this out into `expandTerm` function and document the need as it is confusing institutional knowledge.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+case class SubscriptionUpdate(currentTerm: Option[Int], add: Seq[Add])
+
+object SubscriptionUpdate {
+
+  def holidayCreditToAdd(config: Config)(
+    subscription: Subscription,
+    stoppedPublicationDate: LocalDate
+  ): SubscriptionUpdate = {
+    val credit = Credit.autoRenewingHolidayAmount(subscription)
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: Inline this

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+          )
+        )
+      )
+    )
+  }
+}
+
+case class Add(
+  productRatePlanId: String,
+  contractEffectiveDate: LocalDate,
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: `contractEffectiveDate`, `customerAcceptanceDate`, and `serviceActivationDate` should always be the same so we could require it at construction time.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Zuora.scala [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+object Zuora {
+
+  implicit val backend: SttpBackend[Id, Nothing] = HttpURLConnectionBackend()
+
+  private def normalised[E <: io.circe.Error, R](
+    body: String, result: String => Either[E, R]
+  ): Either[String, R] = result(body).left.map(e => s"Failed to decode '$body': ${e.toString}")
+
+  def subscriptionGetResponse(
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Zuora.scala]: Maybe rename to `getSubscription`?

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        subscription.currentTermPeriodType match {
+          case "Month" => Some(24)
+          case _ => None
+        }
+      } else None
+    SubscriptionUpdate(
+      currentTerm,
+      Seq(
+        Add(
+          productRatePlanId = config.holidayCreditProductRatePlanId,
```
Comment (paulbrown1982) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: This will eventually be product-specific.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala [hunk 18]
```diff
@@ -0,0 +1,16 @@
+package com.gu.holidaystopprocessor
+
+import scala.math.BigDecimal.RoundingMode
+
+object Credit {
+
+  def autoRenewingHolidayAmount(subscription: Subscription): Double = {
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala]: Could we assert price is not wildly out-of-bounds, like say negative million.

File: handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+      termEndDate = LocalDate.now,
+      price = 37.5,
+      billingPeriod = "Quarter",
+      effectiveEndDate = LocalDate.of(2019, 5, 15)
+    )
+    val credit = Credit.autoRenewingHolidayAmount(subscription)
+    credit shouldBe -2.89
+  }
+
+  "autoRenewingHolidayCredit" should "be correct for an annual billing period" in {
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala]: `it` removes having to repeat `autoRenewingHolidayCredit` on every test like so

```
it should "be correct for another quarterly billing period"
```

instead of 

```
"autoRenewingHolidayCredit" should "be correct for another quarterly billing period" 
```

File: handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala [hunk 20]
```diff
@@ -0,0 +1,41 @@
+package com.gu.holidaystopprocessor
+
+import java.time.LocalDate
+
+import org.scalatest.{FlatSpec, Matchers}
+
+class CreditTest extends FlatSpec with Matchers {
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala]: Add test `"autoRenewingHolidayAmount" should "return negative price"`

File: handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/SubscriptionUpdateTest.scala [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+      subscription = Fixtures.mkSubscription(
+        termEndDate = LocalDate.of(2019, 12, 1),
+        price = 42.1,
+        billingPeriod = "Quarter",
+        effectiveEndDate = LocalDate.of(2020, 5, 3)
+      ),
+      stoppedPublicationDate = LocalDate.of(2019, 5, 18)
+    )
+    update shouldBe SubscriptionUpdate(
+      currentTerm = Some(24),
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/SubscriptionUpdateTest.scala]: @paulbrown1982 Can you confirm it is safe to extend `termEndDate` at will?

File: handlers/holiday-stop-processor/cfn.yaml [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+      Code:
+        S3Bucket: support-service-lambdas-dist
+        S3Key: !Sub membership/${Stage}/holiday-stop-processor/holiday-stop-processor.jar
+      Handler: com.gu.holidaystopprocessor.Handler::handle
+      Environment:
+        Variables:
+          Stage: !Ref Stage
+      Role:
+        !GetAtt HolidayStopProcessorRole.Arn
+      MemorySize: 256
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/cfn.yaml]: Put a gig?

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala [hunk 23]
```diff
@@ -0,0 +1,73 @@
+package com.gu.holidaystopprocessor
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: Maybe rename to `com.go.holiday`

File: handlers/holiday-stop-processor/cfn.yaml [hunk 24]
```diff
@@ -0,0 +1,64 @@
+AWSTemplateFormatVersion: "2010-09-09"
+Description: Updates subscriptions with outstanding holiday stops.
```
Comment (mario-galic) [inline handlers/holiday-stop-processor/cfn.yaml]: Add alarm on failed lambda

File: handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/SubscriptionUpdateTest.scala (hunk 21 above)
Comment (paulbrown1982) [reply handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/SubscriptionUpdateTest.scala]: On an auto-renewing sub, yup no problem.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala (hunk 4 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: I've remodelled this slightly in an upcoming PR but will leave this as is here.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala (hunk 5 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: Yes, that would be sensible.  I don't understand why the dev file has a different naming scheme to the other stages in S3.  Shall I just change it?

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala (hunk 6 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala]: Yes, I'll do that.  I was thinking it could be used as a home for all types of credit calculation but that kind of imagining the future is to be avoided.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala (hunk 7 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: Should we be using assertions in production code?  I'd rather set up a scalacheck test of the credit calculation code to ensure it never returns a positive value.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala (hunk 8 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala]: I'll set up a scalacheck test with this as an invariant.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala (hunk 10 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala]: I didn't see that.  Thanks.  Removed now.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/StandaloneApp.scala (hunk 11 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/StandaloneApp.scala]: Done.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Subscription.scala (hunk 12 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Subscription.scala]: Done.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala (hunk 13 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: Done.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala (hunk 14 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: Done.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Zuora.scala (hunk 16 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Zuora.scala]: As discussed, think in a functional language methods should be the name of the thing they return rather than the name of a process as in an OO language.  Although this isn't consistent because the way Scala is used is not entirely one or the other.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala (hunk 17 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: At the moment this code is all very GW specific.  I expect to be parameterised more as its scope expands to more products.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala (hunk 18 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala]: I'll add this to a scalacheck test.

File: handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala (hunk 19 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala]: Done.

File: handlers/holiday-stop-processor/cfn.yaml (hunk 22 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/cfn.yaml]: Shall we try this and see what happens?  In Code it was using less than 200 MB but I'm sure this will increase when the SF bit is integrated.  So I'll expand as needed.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala (hunk 23 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Config.scala]: I was worried that there will be multiple components doing holiday stop work.  An IDE will insert the package name when you create a new class so there should be a small likelihood of a typo.

File: handlers/holiday-stop-processor/cfn.yaml (hunk 24 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/cfn.yaml]: Will do this in a separate PR. https://trello.com/c/DkXszGaJ

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Subscription.scala [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+
+case class Subscription(
+  termEndDate: LocalDate,
+  currentTerm: Int,
+  currentTermPeriodType: String,
+  autoRenew: Boolean,
+  ratePlans: Seq[RatePlan]
+) {
+  val originalRatePlanCharge: Option[RatePlanCharge] = for {
+    ratePlan <- ratePlans.lastOption
```
[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Subscription.scala]: Now sorting by charge effective start date before taking the rate plan.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala (hunk 9 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/HolidayStopProcess.scala]: Same reasoning as below.  ie. functional naming rather than OO naming.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala (hunk 15 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/SubscriptionUpdate.scala]: I think setting them all to the same variable should be enough?

File: handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala (hunk 20 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/test/scala/com/gu/holidaystopprocessor/CreditTest.scala]: Done.

File: handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala (hunk 18 above)
Comment (kelvin-chappell) [reply handlers/holiday-stop-processor/src/main/scala/com/gu/holidaystopprocessor/Credit.scala]: Done.

**Comment by kelvin-chappell** [comment]
@mario-galic:

> * readme documenting architecture and how it relates to Salesforce client    

Card for this: https://trello.com/c/07hEeLaM

> * assurances price is not being increased

There are now some tests for this 

> * cloudwatch alarm via email to SX on failure

Card for this: https://trello.com/c/DkXszGaJ
